# 01. Importing libraries

In [61]:
#importing data
import pandas as pd
import numpy as np
import os

In [62]:
path = r"C:\Users\andyc\Achievement 4 Project"

### 03.)import the orders_products_combined dataframe

In [64]:
df_ords_products_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

### 04.)Check the shape of the imported dataframe

In [66]:
#Comparing the rows and columns with the dataframe beofre it was exported.
df_ords_products_combined.shape

(32434489, 12)

Rows and Columns match the dataframe before it was exported.

In [68]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [69]:
df_prods.shape

(49672, 6)

In [70]:
df_ords_products_combined.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   order_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                  2                  8                     NaN         196   
1                  2                  8                     NaN       14084   
2                  2                  8                     NaN       12427   
3                  2                  8                     NaN       26088   
4                  2                  8                     NaN       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [71]:
#Dropping the _merge column from the previous part of the exercise
df_ords_products_combined = df_ords_products_combined.drop(columns = ['_merge'])

In [72]:
df_ords_products_combined.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   order_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                  2                  8                     NaN         196   
1                  2                  8                     NaN       14084   
2                  2                  8                     NaN       12427   
3                  2                  8                     NaN       26088   
4                  2                  8                     NaN       26405   

   add_to_cart_order  reordered  
0                  1          0  
1                  2          0  
2                  3          0  
3                  4          0  
4                  5          0

Ran df.head to check if the _merge column was deleted and confirmed.

In [73]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

### 05.) Determine a suitable way to combine the orders_products_combined dataframe with your products data set.

In [75]:
#outer join
#When performing an outer merge, it takes results from both dataframes, whether or not they match on the key column.
df_ords_prods_merge_left = df_ords_products_combined.merge(df_prods, on = 'product_id', how = 'left', indicator = True)

In [76]:
df_ords_prods_merge_left

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0                    0             0   2539329        1             1   
1                    0             0   2539329        1             1   
2                    0             0   2539329        1             1   
3                    0             0   2539329        1             1   
4                    0             0   2539329        1             1   
...                ...           ...       ...      ...           ...   
32435331       3421081       3421081   2977660   206209            13   
32435332       3421081       3421081   2977660   206209            13   
32435333       3421081       3421081   2977660   206209            13   
32435334       3421081       3421081   2977660   206209            13   
32435335       3421081       3421081   2977660   206209            13   

          order_day_of_week  order_hour_of_day  days_since_prior_order  \
0                         2                  8                     NaN   
1                         2                  8                     NaN   
2                         2                  8                     NaN   
3                         2                  8                     NaN   
4                         2                  8                     NaN   
...                     ...                ...                     ...   
32435331                  1                 12                     7.0   
32435332                  1                 12                     7.0   
32435333                  1                 12                     7.0   
32435334                  1                 12                     7.0   
32435335                  1                 12                     7.0   

          product_id  add_to_cart_order  reordered  Unnamed: 0_y  \
0                196                  1          0         195.0   
1              14084                  2          0       14084.0   
2              12427                  3          0       12427.0   
3              26088                  4          0       26089.0   
4              26405                  5          0       26406.0   
...              ...                ...        ...           ...   
32435331       14197                  5          1       14197.0   
32435332       38730                  6          0       38734.0   
32435333       31477                  7          0       31479.0   
32435334        6567                  8          0        6567.0   
32435335       22920                  9          0       22921.0   

                                     product_name  aisle_id  department_id  \
0                                            Soda      77.0            7.0   
1         Organic Unsweetened Vanilla Almond Milk      91.0           16.0   
2                             Original Beef Jerky      23.0           19.0   
3                      Aged White Cheddar Popcorn      23.0           19.0   
4                XL Pick-A-Size Paper Towel Rolls      54.0           17.0   
...                                           ...       ...            ...   
32435331                             Tomato Paste       9.0            9.0   
32435332          Brownie Crunch High Protein Bar       3.0           19.0   
32435333    High Protein Bar Chunky Peanut Butter       3.0           19.0   
32435334      Chocolate Peanut Butter Protein Bar       3.0           19.0   
32435335      Roasted & Salted Shelled Pistachios     117.0           19.0   

          prices _merge  
0            9.0   both  
1           12.5   both  
2            4.4   both  
3            4.7   both  
4            1.0   both  
...          ...    ...  
32435331     5.6   both  
32435332     5.9   both  
32435333     4.2   both  
32435334     4.9   both  
32435335     1.7   both  

[32435336 rows x 17 columns]

Above are the results of the outer merge.

In [78]:
#Using the value_counts() function lets you quickly sum up all the values in the “_merge” column, letting you see instantly whether you have a full match or not.
df_ords_prods_merge_left['_merge'].value_counts()

_merge
both          32405122
left_only        30214
right_only           0
Name: count, dtype: int64

Above are the results for the value counts of the outer merge. We can see that there are both values in the both and left_only.

In [80]:
# This will create a new dataframe, df_merged_large, that contains the combined df_ords and df_ords_prior dataframes and uses the “order_id” column as its key. 
# Only the rows that match on product_id in both dataframes will be kept/included in the resulting dataframe.
df_ords_prods_merge = df_prods.merge(df_ords_products_combined, on = 'product_id', indicator = 'merge_')

In [81]:
df_ords_prods_merge

Unnamed: 0_x  product_id                product_name  aisle_id  \
0                    0           1  Chocolate Sandwich Cookies        61   
1                    0           1  Chocolate Sandwich Cookies        61   
2                    0           1  Chocolate Sandwich Cookies        61   
3                    0           1  Chocolate Sandwich Cookies        61   
4                    0           1  Chocolate Sandwich Cookies        61   
...                ...         ...                         ...       ...   
32405117         49692       49688      Fresh Foaming Cleanser        73   
32405118         49692       49688      Fresh Foaming Cleanser        73   
32405119         49692       49688      Fresh Foaming Cleanser        73   
32405120         49692       49688      Fresh Foaming Cleanser        73   
32405121         49692       49688      Fresh Foaming Cleanser        73   

          department_id  prices  Unnamed: 0.1  Unnamed: 0_y  order_id  \
0                    19     5.8          1987          1987   3139998   
1                    19     5.8          1989          1989   1977647   
2                    19     5.8         11433         11433    389851   
3                    19     5.8         12198         12198    652770   
4                    19     5.8         12200         12200   1813452   
...                 ...     ...           ...           ...       ...   
32405117             11    13.5       3321928       3321928   1788356   
32405118             11    13.5       3324686       3324686   3401313   
32405119             11    13.5       3332310       3332310    809510   
32405120             11    13.5       3332314       3332314   2359893   
32405121             11    13.5       3416620       3416620   2385091   

          user_id  order_number  order_day_of_week  order_hour_of_day  \
0             138            28                  6                 11   
1             138            30                  6                 17   
2             709             2                  0                 21   
3             764             1                  3                 13   
4             764             3                  4                 17   
...           ...           ...                ...                ...   
32405117   200215             2                  0                  9   
32405118   200377             1                  4                 11   
32405119   200873             5                  3                  8   
32405120   200873             9                  3                 15   
32405121   205926            11                  1                 15   

          days_since_prior_order  add_to_cart_order  reordered merge_  
0                            3.0                  5          0   both  
1                           20.0                  1          1   both  
2                            6.0                 20          0   both  
3                            NaN                 10          0   both  
4                            9.0                 11          1   both  
...                          ...                ...        ...    ...  
32405117                     5.0                 27          0   both  
32405118                     NaN                  5          0   both  
32405119                    15.0                 12          0   both  
32405120                     5.0                 11          1   both  
32405121                     6.0                 35          0   both  

[32405122 rows x 17 columns]

Above is the result for the default inner join.

I ran both an outer join and inner join concluded that both of these data sets have a full merge rate. The left and inner join had the same number of rows.

### 06.) Confirm the results of the merge using the merge flag.

In [85]:
#Using the value_counts() function lets you quickly sum up all the values in the “_merge” column, letting you see instantly whether you have a full match or not.
df_ords_prods_merge['merge_'].value_counts()

merge_
both          32405122
left_only            0
right_only           0
Name: count, dtype: int64

Comparing the left join and inner join, we can see that the Both columns in both joins have the exact same value. The left_only has 30214 which represents the missing values from the left join.

In [87]:
#creates a subset that provides all the missing values
df_missing_prices_values = df_ords_prods_merge_left[df_ords_prods_merge_left['prices'].isnull() == True]

In [88]:
df_missing_prices_values

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
2963               376           376      7099       27            63   
3205               393           393   1837192       27            80   
3253               398           398   3331846       28             3   
3395               416           416   1385910       28            21   
4694               558           558   2873174       38             1   
...                ...           ...       ...      ...           ...   
32425571       3419963       3419963    443233   206139             2   
32425577       3419964       3419964   1507543   206139             3   
32427375       3420197       3420197   1282274   206155             4   
32432183       3420758       3420758   1586626   206194             9   
32432206       3420760       3420760     56974   206194            11   

          order_day_of_week  order_hour_of_day  days_since_prior_order  \
2963                      3                 10                     1.0   
3205                      2                  8                     6.0   
3253                      0                 16                     2.0   
3395                      2                 17                     1.0   
4694                      0                 10                     NaN   
...                     ...                ...                     ...   
32425571                  1                 17                    17.0   
32425577                  1                  8                    14.0   
32427375                  6                  9                     3.0   
32432183                  1                 20                    17.0   
32432206                  5                  0                    30.0   

          product_id  add_to_cart_order  reordered  Unnamed: 0_y product_name  \
2963            6799                  1          0           NaN          NaN   
3205            6799                  9          1           NaN          NaN   
3253              34                  6          0           NaN          NaN   
3395              34                  1          1           NaN          NaN   
4694             116                 11          0           NaN          NaN   
...              ...                ...        ...           ...          ...   
32425571          34                  2          1           NaN          NaN   
32425577          34                  2          1           NaN          NaN   
32427375        2240                 18          0           NaN          NaN   
32432183        1511                  6          0           NaN          NaN   
32432206        1511                  9          1           NaN          NaN   

          aisle_id  department_id  prices     _merge  
2963           NaN            NaN     NaN  left_only  
3205           NaN            NaN     NaN  left_only  
3253           NaN            NaN     NaN  left_only  
3395           NaN            NaN     NaN  left_only  
4694           NaN            NaN     NaN  left_only  
...            ...            ...     ...        ...  
32425571       NaN            NaN     NaN  left_only  
32425577       NaN            NaN     NaN  left_only  
32427375       NaN            NaN     NaN  left_only  
32432183       NaN            NaN     NaN  left_only  
32432206       NaN            NaN     NaN  left_only  

[30214 rows x 17 columns]

We compare the missing values to the left only from the left join and we can conclude that there are 30214 rows that have NULL values according to the left join.

### 07.)Export the newly created dataframe as ords_prods_merge in a suitable format

In [91]:
#exporting as a pickle file
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))

We export the merge without prices values that contain NULL values.